source/simulator/core.cc

```diff
@@ -755,20 +778,75 @@ namespace aspect
                                                             c),
                                                             introspection.component_masks.compositional_fields[c].first_selected_component(),
                                                             introspection.n_components),
-                                                        current_constraints,
+                                                        new_current_constraints,
                                                         introspection.component_masks.compositional_fields[c]);
             }
       }
 
+    if (!boundary_composition_manager.allows_fixed_composition_on_outflow_boundaries())
+      restore_outflow_boundary_ids(boundary_id_offset);
+
     // let plugins add more constraints if they so choose, then close the
     // constraints object
-    signals.post_constraints_creation(*this, current_constraints);
+    signals.post_constraints_creation(*this, new_current_constraints);
 
-    current_constraints.close();
+    new_current_constraints.close();
 
     // let the melt handler add its constraints once before we solve the porosity system for the first time
     if (parameters.include_melt_transport)
-      melt_handler->save_constraints (current_constraints);
+      melt_handler->save_constraints (new_current_constraints);
+
+    // Now check if the current_constraints we just computed changed from before. Do this before the melt handler
+    // adds constraints for the melt cells, because they are allowed to change (and often do) without us having
+    // to reinit the sparsity pattern.
+    // If the mesh got refined and the size of the linear system changed, the old and new constraint
+    // matrices will have different entries, and we can not easily compare them.
+    // However, in case of mesh refinement we have to rebuild the matrix anyway, so we can skip the
+    // step that checks if constraints have changed.
```

Comment:
couldn't you do this easier by invalidating the `current_constraints` object at the time of mesh refinement?

Reply:
@tjhei designed this part and can answer that question much better than I can.

Reply:
Let's not worry about it then.